frame/referenda/src/lib.rs

```diff
@@ -1283,4 +1296,90 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			Self::deposit_event(Event::<T, I>::MetadataCleared { index, hash });
 		}
 	}
+
+	/// Ensure the correctness of the state of this pallet.
+	///
+	/// The following assertions must always apply.
+	///
+	/// General assertions:
+	///
+	/// * `ReferendumCount` must always be equal to the number of referenda in `ReferendumInfoFor`.
+	/// * Referendum indices in `MetadataOf` must also be stored in `ReferendumInfoFor`.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), TryRuntimeError> {
+		ensure!(
+			ReferendumCount::<T, I>::get() as usize ==
+				ReferendumInfoFor::<T, I>::iter_keys().count(),
+			"Number of referenda in `ReferendumInfoFor` is different than `ReferendumCount`"
+		);
+
+		MetadataOf::<T, I>::iter_keys().try_for_each(|referendum_index| -> DispatchResult {
+			ensure!(
+				ReferendumInfoFor::<T, I>::contains_key(referendum_index),
+				"Referendum indices in `MetadataOf` must also be stored in `ReferendumInfoOf`"
+			);
+			Ok(())
+		})?;
+
+		Self::try_state_referenda_info()?;
+		Self::try_state_tracks()?;
+
+		Ok(())
+	}
+
+	/// Looking at referenda info:
+	///
+	/// - Data regarding ongoing phase:
+	///
+	/// * There must exist track info for the track of the referendum.
+	/// * The deciding stage has to begin before confirmation period.
+	/// * If alarm is set the nudge call has to be at most `UndecidingTimeout` blocks away
+	///  from the submission block.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn try_state_referenda_info() -> Result<(), TryRuntimeError> {
+		ReferendumInfoFor::<T, I>::iter().try_for_each(|(_, referendum)| {
+			match referendum {
+				ReferendumInfo::Ongoing(status) => {
+					ensure!(
+						Self::track(status.track).is_some(),
+						"No track info for the track of the referendum."
+					);
+
+					if let Some(deciding) = status.deciding {
+						ensure!(
+							deciding.since <
+								deciding.confirming.unwrap_or(T::BlockNumber::max_value()),
+								"Deciding status cannot begin before confirming stage."
+						)
+					}
+
+					if let Some(alarm) = status.alarm {
+						ensure!(alarm.0 <= status.submitted.saturating_add(T::UndecidingTimeout::get()), "The alarm cannot be set more than `UndecidingTimeout` away from the submission block.")
+					}
```

Comment:
```suggestion
```
This is true for the first alarm, but the alarm can be set after the first nudge call(e.g. at the beginning of the deciding period). For this reason, this check has been failing against the Polkadot state and I am removing it.